tests/test_optimizer/test_dist_adafactor.py

```diff
@@ -0,0 +1,719 @@
+            shard_to_param=shard_to_param,
+            use_zero=use_zero,
+        )
+    else:
+        shard_to_param = set_master_param_to_shard_param(tp_param_group)
+        dist_optim.setup_distributed(
+            tensor_parallel_group=tp_group,
+            data_parallel_group=dp_group,
+            shard_to_param=shard_to_param,
+            use_zero=use_zero,
+        )
+
+    # ==============================
+    # Booster Init
+    # ==============================
+    plugin = TorchDDPPlugin()
+    booster = Booster(plugin=plugin)
+    criterion = lambda x: x.mean()
+
+    tp_model, dist_optim, criterion, _, _ = booster.boost(tp_model, dist_optim, criterion)
+
+    # ==============================
+    # Correctness Verify
+    # ==============================
+    x = torch.randn(HEIGHT, WIDTH, device=local_rank)
+
+    out = base_model(x)
+    out_tp = tp_model(x)
+
+    if zero_size > 1:
+        dist_optim.backward(out_tp.sum())
+        base_optim.backward(out.sum())
+    else:
+        out_tp.sum().backward()
+        out.sum().backward()
+
+    base_optim.step()
+    dist_optim.step()
+
+    base_optim.zero_grad()
+    dist_optim.zero_grad()
+    for p, tp_p in zip(base_param_group, tp_param_group):
+        param_is_distributed = is_distributed_tensor(tp_p)
+        if param_is_distributed:
+            shard_spec = get_sharding_spec(tp_p)
+            if len(shard_spec.sharding_sequence) >= 2:
+                # Col Parallel
+                if shard_spec.sharding_sequence[0] == "R":
+                    tp_p = _gather(input_=tp_p, dim=-1, process_group=tp_group)  # gather
+                # ROW Parallel
+                if shard_spec.sharding_sequence[-1] == "R":
+                    tp_p = _gather(input_=tp_p, dim=0, process_group=tp_group)  # gather
+            else:
+                # TP bias
+                tp_p = _gather(input_=tp_p, dim=-1, process_group=tp_group)  # gather
+
+        else:
+            # No TP bias
+            pass
+        correctness_verify(p.data, tp_p.data, dtype)
+
+
+@parameterize(
+    "test_config",
+    [
+        {
+            "tp_size": 1,
+            "num_microbatches": 4,
+            "zero_stage": 2,
+            "precision": "bf16",
+        },
+        {
+            "tp_size": 2,
+            "num_microbatches": 4,
+            "zero_stage": 2,
+            "precision": "bf16",
+        },
+        {
+            "tp_size": 4,
+            "num_microbatches": 4,
+            "zero_stage": 2,
+            "precision": "bf16",
+        },
+        {
+            "tp_size": 1,
+            "num_microbatches": 4,
+            "zero_stage": 2,
+            "precision": "fp16",
+        },
+        {
+            "tp_size": 2,
+            "num_microbatches": 4,
+            "zero_stage": 2,
+            "precision": "fp16",
+        },
+        {
+            "tp_size": 4,
+            "num_microbatches": 4,
+            "zero_stage": 2,
+            "precision": "fp16",
```

Comment:
Only testing bf16 is OK

Reply:
> Only testing bf16 is OK


Fixed in https://github.com/hpcaitech/ColossalAI/pull/5484/commits/150ac1975f7e472b661373abd9b13036e4612514;